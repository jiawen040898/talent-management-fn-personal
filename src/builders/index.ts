export * from './base-query.builder';
export * from './employee-task-query.builder';
export * from './feedback-cycle-reviewee-query.builder';
export * from './feedback-cycle-reviewer-query.builder';
export * from './learning-goal-query.builder';
