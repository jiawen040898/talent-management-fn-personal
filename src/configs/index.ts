export * from './aws.config';
export * from './gcp.config';
export * from './recommendation.config';
