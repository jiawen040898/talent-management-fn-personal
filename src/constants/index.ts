export * from './action-source.enum';
export * from './big-queries.enum';
export * from './cache-store.enum';
export * from './company-module.enum';
export * from './company-status.enum';
export * from './due-type.enum';
export * from './employee-score.enum';
export * from './employee-task.enum';
export * from './event-type.enum';
export * from './feedback-recommendation.enum';
export * from './fit-score-recipe.enum';
export * from './person-score.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework-type.enum';
export * from './sqs-event-type.enum';
export * from './unleash-errors.enum';
