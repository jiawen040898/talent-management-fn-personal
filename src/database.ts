import { secretService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ENTITIES } from './models';

type DbSecret = {
    username: string;
    engine: string;
    port: number;
    host: string;
    password: string;
    schema: string;
    dbname: string;
    dbInstanceIdentifier: string;
};

let dataSource: DataSource;

export const getDataSource = async (): Promise<DataSource> => {
    if (dataSource?.isInitialized) {
        return dataSource;
    }

    const dbSecret = await secretService.getSecret<DbSecret>(
        process.env.SM_NAME as string,
    );

    dataSource = await new DataSource({
        type: 'postgres',
        host: dbSecret.host,
        port: dbSecret.port,
        username: dbSecret.username,
        password: dbSecret.password,
        database: dbSecret.dbname,
        schema: dbSecret.schema,
        entities: [...ENTITIES],
        maxQueryExecutionTime: 1000,
        synchronize: false,
        migrationsRun: false,
        migrations: [],
        extra: {
            connectionTimeoutMillis: 3000,
            idleTimeoutMillis: 1000,
            max: 1,
        },
        logging: ['error'],
    }).initialize();

    return dataSource;
};
