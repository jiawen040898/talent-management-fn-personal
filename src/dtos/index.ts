export * from './company.dto';
export * from './cron-schedule-message.dto';
export * from './employee.dto';
export * from './employee-account-created-event.dto';
export * from './employee-assessment-score-event.dto';
export * from './employee-fit-score-event.dto';
export * from './feedback-cycle.dto';
export * from './fit-score-recipe.dto';
export * from './manager-subordinate-aggregated-report-data.dto';
