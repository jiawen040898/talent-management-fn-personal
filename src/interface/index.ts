export * from '../dtos/fit-score-recipe.dto';
export * from './cognitive-score.interface';
export * from './employee-score.interface';
export * from './event-model.interface';
export * from './feedback-cycle.interface';
export * from './feedback-recommendation.interface';
export * from './fit-score.interface';
export * from './personality-score.interface';
export * from './program.interface';
export * from './redis.interface';
