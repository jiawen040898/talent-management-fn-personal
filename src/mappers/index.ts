export * from './employee-score.mapper';
export * from './feedback.mapper';
export * from './feedback-recommendation.mapper';
