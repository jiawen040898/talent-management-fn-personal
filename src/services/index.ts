export * from './cognitive-score.service';
export * from './company.service';
export * from './domain-event-router.service';
export * from './employee-daily-reminder.service';
export * from './employee-management.service';
export * from './employee-progress-report.service';
export * from './employee-score.service';
export * from './feedback-cycle-action.service';
export * from './feedback-cycle-dashboard-provision.service';
export * from './feedback-cycle-reminder.service';
export * from './feedback-cycle-reviewee-recommendation.service';
export * from './fit-score.service';
export * from './personality-score.service';
export * from './psychology.service';
export * from './recommendation.service';
