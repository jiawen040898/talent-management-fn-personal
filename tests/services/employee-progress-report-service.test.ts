import { SNSClient } from '@aws-sdk/client-sns';
import { SQSClient } from '@aws-sdk/client-sqs';
import { GcpBigQueryService } from '@pulsifi/fn/services/gcp-big-query';

import { EmployeeProgressReportService } from '../../src/services/';
import { testData } from './fixtures/test-data';

const mockQuery = jest.spyOn(GcpBigQueryService, 'query');

const spySNSClient = jest
    .spyOn(SNSClient.prototype, 'send')
    .mockImplementation(jest.fn());

const spySQSClient = jest
    .spyOn(SQSClient.prototype, 'send')
    .mockImplementation();

describe('EmployeeProgressReportService', () => {
    let sut: EmployeeProgressReportService;
    beforeAll(async () => {
        sut = new EmployeeProgressReportService();
    });

    beforeEach(() => {
        spySQSClient.mockClear();
        spySNSClient.mockClear();
        mockQuery.mockClear();
    });

    describe('handleEmployeeProgressReportFromCronJob', () => {
        it('should send a list of sqs request per manager from company', async () => {
            // Arrange
            mockQuery.mockResolvedValue(testData.mockBQManagerAndStat);

            // Act
            await sut.handleEmployeeProgressReportFromCronJob([
                testData.mockCompanyLookupData,
            ]);

            // Assert
            const input: SafeAny = spySQSClient.mock.calls[0][0]?.input;
            const content: SafeAny[] = input?.Entries.map(
                // get content from batchSQSCommand
                (sqsCommand: SafeAny) => JSON.parse(sqsCommand.MessageBody),
            );

            expect({ content }).toMatchSnapshot({
                content: new Array(content.length).fill({
                    event_id: expect.any(String),
                }),
            });
            expect(mockQuery.mock.calls).toMatchSnapshot();
        });

        it('should not send anything nor prompt error if non of the company match the module', async () => {
            // Act
            await sut.handleEmployeeProgressReportFromCronJob([
                testData.mockCompanyLookupDataWithoutModule,
            ]);

            // Assert
            expect(mockQuery).not.toBeCalled();
            expect(spySQSClient).not.toBeCalled();
        });
    });

    describe('handleEmployeeProgressReportByManager', () => {
        it('should send report to notification queue', async () => {
            // Arrange
            mockQuery.mockResolvedValue(testData.mockSubordinateSummaries);

            // Act
            await sut.handleEmployeeProgressReportByManager(
                testData.mockManagerReportData.company_id,
                testData.mockManagerReportData,
            );

            // Assert
            expect(spySNSClient).toHaveBeenCalledTimes(1);
            expect(mockQuery.mock.calls).toMatchSnapshot();
        });

        it('should not send anything if there is no', async () => {
            // Arrange
            mockQuery.mockResolvedValue([]);

            // Act
            await sut.handleEmployeeProgressReportByManager(
                testData.mockManagerReportData.company_id,
                testData.mockManagerReportData,
            );

            // Assert
            expect(spySNSClient).toHaveBeenCalledTimes(0);
        });
    });
});
