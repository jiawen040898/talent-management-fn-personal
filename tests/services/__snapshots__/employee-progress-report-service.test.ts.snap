// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeProgressReportService handleEmployeeProgressReportByManager should send report to notification queue 1`] = `
[
  [
    {
      "clientLibraryConfig": {},
      "projectId": "data-sandbox-warehouse",
      "region": "asia-southeast1",
    },
    "
        SELECT employee_id, 
          name, 
          (IFNULL(alltime_goal_overdue, 0) + IFNULL(alltime_action_overdue, 0)) AS overdue_count, 
          (IFNULL(weekly_number_of_new_goal_created, 0) + IFNULL(weekly_number_of_new_action_item_created, 0)) AS new_count,
          (IFNULL(weekly_goal_completed, 0) + IFNULL(weekly_action_completed, 0)) AS completed_count
          FROM \`{{project_id}}.product_dept_talent_mgmt{{region_postfix}}.lxp_email_notification\`
          WHERE
          company_id = @company_id
            AND manager_employee_id = @manager_employee_id
            AND is_deleted = false
          ORDER BY overdue_count DESC, name ASC
          LIMIT 5
    ",
    {
      "company_id": 5,
      "manager_employee_id": "c77021a7-cb4f-466c-9f51-b7cd73eaeecf",
    },
  ],
]
`;

exports[`EmployeeProgressReportService handleEmployeeProgressReportFromCronJob should send a list of sqs request per manager from company 1`] = `
{
  "content": [
    {
      "company_id": 5,
      "data": {
        "company_id": 5,
        "company_is_live": false,
        "company_name": "Bitcoin Incorporation",
        "manager_email": "joshborbon@getnada.com",
        "manager_employee_id": "dd394bf0-5d0d-44ce-adf4-6e500417f645",
        "manager_first_name": "Josh",
        "manager_name": "Josh Borbon",
        "manager_user_account_id": 11973,
        "total_action_completed": 16,
        "total_action_overdue": 12,
        "total_goal_completed": 9,
        "total_goal_overdue": 6,
        "total_new_action": 9,
        "total_new_goal": 14,
      },
      "event_id": Any<String>,
      "event_type": "talent_management_handling_manager_weekly_progress",
      "user_account_id": 0,
    },
    {
      "company_id": 5,
      "data": {
        "company_id": 5,
        "company_is_live": false,
        "company_name": "Bitcoin Incorporation",
        "manager_email": "greatmanager@getnada.com",
        "manager_employee_id": "c863d0fa-7fc7-4f88-96cd-9d0405224835",
        "manager_first_name": "Jay",
        "manager_name": "Jay Manager",
        "manager_user_account_id": 13282,
        "total_action_completed": 0,
        "total_action_overdue": 10,
        "total_goal_completed": 0,
        "total_goal_overdue": 11,
        "total_new_action": 4,
        "total_new_goal": 6,
      },
      "event_id": Any<String>,
      "event_type": "talent_management_handling_manager_weekly_progress",
      "user_account_id": 0,
    },
    {
      "company_id": 5,
      "data": {
        "company_id": 5,
        "company_is_live": false,
        "company_name": "Bitcoin Incorporation",
        "manager_email": "pulsifimanager01@getnada.com",
        "manager_employee_id": "eae5cf89-996c-43e8-b9e2-edd72e989978",
        "manager_first_name": "Anna",
        "manager_name": "Anna Hong",
        "manager_user_account_id": 12181,
        "total_action_completed": 2,
        "total_action_overdue": 5,
        "total_goal_completed": 0,
        "total_goal_overdue": 2,
        "total_new_action": 4,
        "total_new_goal": 3,
      },
      "event_id": Any<String>,
      "event_type": "talent_management_handling_manager_weekly_progress",
      "user_account_id": 0,
    },
  ],
}
`;

exports[`EmployeeProgressReportService handleEmployeeProgressReportFromCronJob should send a list of sqs request per manager from company 2`] = `
[
  [
    {
      "clientLibraryConfig": {},
      "projectId": "data-sandbox-warehouse",
      "region": "asia-southeast1",
    },
    "
        SELECT max(company_id) AS company_id, 
          max(company_name) AS company_name, 
          max(company_is_live) AS company_is_live,
          max(manager_name) AS manager_name, manager_employee_id,
          max(manager_first_name) AS manager_first_name,
          max(manager_user_account_id) AS manager_user_account_id,
          max(manager_email) AS manager_email,
          IFNULL(sum(weekly_action_completed), 0) AS total_action_completed, 
          IFNULL(sum(alltime_action_overdue), 0) AS total_action_overdue, 
          IFNULL(sum(weekly_number_of_new_action_item_created), 0) AS total_new_action,
          IFNULL(sum(weekly_goal_completed), 0) AS total_goal_completed, 
          IFNULL(sum(alltime_goal_overdue), 0) AS total_goal_overdue, 
          IFNULL(sum(weekly_number_of_new_goal_created), 0) AS total_new_goal,
          FROM \`{{project_id}}.product_dept_talent_mgmt{{region_postfix}}.lxp_email_notification\` 
          WHERE company_id IN UNNEST(@company_id)
            AND is_deleted = false
          GROUP BY  manager_employee_id
    ",
    {
      "company_id": [
        5,
      ],
    },
  ],
]
`;
